src/native/libs/System.Security.Cryptography.Native/pal_evp.c

```diff
@@ -10,31 +10,55 @@
 
 #define SUCCESS 1
 
-static const EVP_MD* g_evpFetchMd5 = NULL;
-static pthread_once_t g_evpFetch = PTHREAD_ONCE_INIT;
-
-static void EnsureFetchEvpMdAlgorithms(void)
-{
-    // This is called from a pthread_once - this method should not be called directly.
-
 #ifdef NEED_OPENSSL_3_0
-    if (API_EXISTS(EVP_MD_fetch))
-    {
-        ERR_clear_error();
-
-        // Try to fetch an MD5 implementation that will work regardless if
-        // FIPS is enforced or not.
-        g_evpFetchMd5 = EVP_MD_fetch(NULL, "MD5", "-fips");
+#define SETUP_MD_FETCH(export, fn, name, query) \
+    static const EVP_MD* g_evpFetch##export = NULL; \
+    static pthread_once_t g_evpFetchInit##export = PTHREAD_ONCE_INIT; \
+    static void EnsureFetchEvpMd##export(void) \
+    { \
+        if (API_EXISTS(EVP_MD_fetch)) \
+        { \
+            ERR_clear_error(); \
+            g_evpFetch##export = EVP_MD_fetch(NULL, name, query); \
+        } \
+\
+        if (g_evpFetch##export == NULL) \
+        { \
+            g_evpFetch##export = fn(); \
+        } \
+    } \
+    \
+    const EVP_MD* CryptoNative_##export(void) \
+    { \
+        pthread_once(&g_evpFetchInit##export, EnsureFetchEvpMd##export); \
+        return g_evpFetch##export; \
     }
+#define SETUP_MD_FETCH_LIGHTUP_SHA3(...) SETUP_MD_FETCH(__VA_ARGS__)
+#else
+#define SETUP_MD_FETCH(export, fn, name, query) \
+    const EVP_MD* CryptoNative_##export(void) \
+    { \
+        return fn(); \
+    }
+#if HAVE_OPENSSL_SHA3
```

Comment:
We can't have `#if` inside of a macro so we need to invert it and define the macro differently depending on the `#if`. `HAVE_OPENSSL_SHA3` is `1` if: they are a non-portable build and we found our SHA-3 functions from cmake config. OR, if they are a portable build, we flip HAVE_OPENSSL_SHA3 to `1` and rely on a runtime lookup, which is what API_EXISTS is for.